File: snippets/design-patterns/creational/builder.py
'''
--- Builder Pattern ---
Use Case: When you need to construct complex objects step by step.
Example: Creating objects with many optional parameters.
'''

class Car:
    def __init__(self):
        self.make = None
        self.model = None
        self.color = None

    def __str__(self):
        return f"{self.color} {self.make} {self.model}"

class CarBuilder:
    def __init__(self):
        self.car = Car()

    def set_make(self, make):
        self.car.make = make
        return self

    def set_model(self, model):
        self.car.model = model
        return self

    def set_color(self, color):
        self.car.color = color
        return self

    def build(self):
        return self.car

# Example Usage
builder = CarBuilder()
car = builder.set_make("Tesla").set_model("Model S").set_color("Red").build()
print(car)  # Output: Red Tesla Model S

File: snippets/design-patterns/creational/builder.go
// --- Builder Pattern ---
// Use Case: Construct complex objects step by step.
// Example: Building objects with many optional parameters.
// Description: The Builder Pattern separates the construction of a complex object
// from its representation, allowing the same construction process to create different representations.

package main

import "fmt"

type Car struct {
	Make  string
	Model string
	Color string
}

type CarBuilder struct {
	make  string
	model string
	color string
}

func (cb *CarBuilder) SetMake(make string) *CarBuilder {
	cb.make = make
	return cb
}

func (cb *CarBuilder) SetModel(model string) *CarBuilder {
	cb.model = model
	return cb
}

func (cb *CarBuilder) SetColor(color string) *CarBuilder {
	cb.color = color
	return cb
}

func (cb *CarBuilder) Build() Car {
	return Car{
		Make:  cb.make,
		Model: cb.model,
		Color: cb.color,
	}
}

func main() {
	car := (&CarBuilder{}).SetMake("Tesla").SetModel("Model S").SetColor("Red").Build()
	fmt.Printf("Car: %+v\n", car) // Output: Car: {Make:Tesla Model:Model S Color:Red}
}
